# Data Preprocessing

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.gridspec as gridspec
from collections import Counter
import warnings
warnings.filterwarnings("ignore")

In [3]:
def load_data(data_file):
    users, items = [], []
    interactions_per_user = {}

    with open (data_file) as f:
        for line in f:
            parts = line.strip().split()
            if len(parts) < 2:
                continue
            user_id = int(parts[0])
            item_ids = [int(x) for x in parts[1:]]
            interactions_per_user[user_id] = item_ids
            users.extend([user_id] * len(item_ids))
            items.extend(item_ids)
    return np.array(users), np.array(items), interactions_per_user

def analyse_dataset(current_dataset):
    train_users, train_items, train_dict = load_data(f"{current_dataset}/train.txt")
    test_users, test_items, test_dict = load_data(f"{current_dataset}/test.txt")

    user_list_path = f"{current_dataset}/user_list.txt"
    item_list_path = f"{current_dataset}/item_list.txt"
    
    if user_list_path:
        user_list = pd.read_csv(f"{current_dataset}/user_list.txt", sep =" ")
    if item_list_path:
        item_list = pd.read_csv(f"{current_dataset}/item_list.txt", sep =" ")
    

    n_users = len(user_list)
    n_items = len(item_list)
    n_train = len(train_users)
    n_test = len(test_users)
    train_users_set = set(train_users)
    test_users_set  = set(test_users)
    train_items_set = set(train_items)
    test_items_set  = set(test_items)

    density_train = n_train / (n_users * n_items)
    density_test  = n_test  / (n_users * n_items)

    print(f"{current_dataset} dataset overview")
    print(f"Total users (user_list)   : {n_users:,}")
    print(f"Total items (item_list)   : {n_items:,}")
    print(f"Train interactions        : {n_train:,}")
    print(f"Test  interactions        : {n_test:,}")
    print(f"Train+Test interactions   : {n_train + n_test:,}")
    print(f"Train sparsity            : {density_train:.6f}  ({density_train*100:.6f}%)")
    print(f"Test  sparsity            : {density_test:.6f}  ({density_test*100:.6f}%)")
    
    print("\nUser Coverage")
    print(f"Users in train            : {len(train_users_set):,} / {n_users:,} ({len(train_users_set)/n_users*100:.3f}%)")
    print(f"Users in test             : {len(test_users_set):,} / {n_users:,} ({len(test_users_set)/n_users*100:.3f}%)")

    
    print("\nItem Coverage")
    print(f"Items in train: {len(train_items_set):,} / {n_items:,} ({len(train_items_set)/n_items*100:.3f}%)")
    print(f"Items in test: {len(test_items_set):,} / {n_items:,} ({len(test_items_set)/n_items*100:.3f}%)")
    
    train_counts = np.array([len(v) for v in train_dict.values()])
    test_counts  = np.array([len(v) for v in test_dict.values()])
    print("\nInteractions per User Train and Test")
    print(f"  Mean : {train_counts.mean():.2f}, {test_counts.mean():.2f}")
    print(f"  Median  : {np.median(train_counts):.1f}, {np.median(test_counts):.1f}")
    print(f"  Std     : {train_counts.std():.2f}, {test_counts.std():.2f}")
    print(f"  Min     : {train_counts.min()}, {test_counts.min()}")
    print(f"  Max     : {train_counts.max()}, {test_counts.max()}")


    #4. Popularity
    train_item_counts = Counter(train_items)
    test_item_counts  = Counter(test_items)
    train_pop = np.array(list(train_item_counts.values()))
    test_pop  = np.array(list(test_item_counts.values()))
    
    print(f"Mean interactions per item  : {train_pop.mean():.5f}")
    print(f"Median : {np.median(train_pop):.5f}")
    print(f"Max (most popular) : {train_pop.max()}")
    topk_items = train_item_counts.most_common(20)
    print(f"  Top-20 items (id, count) : {topk_items}")



**Yelp2018**

In [4]:
analyse_dataset("data/yelp2018")

data/yelp2018 dataset overview
Total users (user_list)   : 31,668
Total items (item_list)   : 38,048
Train interactions        : 1,237,259
Test  interactions        : 324,147
Train+Test interactions   : 1,561,406
Train sparsity            : 0.001027  (0.102685%)
Test  sparsity            : 0.000269  (0.026902%)

User Coverage
Users in train            : 31,668 / 31,668 (100.000%)
Users in test             : 31,668 / 31,668 (100.000%)

Item Coverage
Items in train: 38,048 / 38,048 (100.000%)
Items in test: 36,073 / 38,048 (94.809%)

Interactions per User Train and Test
  Mean : 39.07, 10.24
  Median  : 25.0, 7.0
  Std     : 45.11, 11.25
  Min     : 16, 2
  Max     : 1848, 463
Mean interactions per item  : 32.51837
Median : 17.00000
Max (most popular) : 1258
  Top-20 items (id, count) : [(783, 1258), (786, 1147), (3245, 1061), (2902, 1037), (2014, 1014), (773, 905), (3647, 865), (2835, 833), (2012, 821), (710, 797), (852, 753), (13, 745), (15, 720), (3651, 710), (657, 707), (1031, 696), 

**Gowalla**

In [5]:
analyse_dataset("data/gowalla")

data/gowalla dataset overview
Total users (user_list)   : 29,858
Total items (item_list)   : 40,981
Train interactions        : 810,128
Test  interactions        : 217,242
Train+Test interactions   : 1,027,370
Train sparsity            : 0.000662  (0.066208%)
Test  sparsity            : 0.000178  (0.017754%)

User Coverage
Users in train            : 29,858 / 29,858 (100.000%)
Users in test             : 29,858 / 29,858 (100.000%)

Item Coverage
Items in train: 40,981 / 40,981 (100.000%)
Items in test: 38,546 / 40,981 (94.058%)

Interactions per User Train and Test
  Mean : 27.13, 7.28
  Median  : 16.0, 4.0
  Std     : 36.86, 9.22
  Min     : 8, 1
  Max     : 811, 203
Mean interactions per item  : 19.76838
Median : 12.00000
Max (most popular) : 1415
  Top-20 items (id, count) : [(2525, 1415), (21536, 1329), (559, 1321), (192, 1190), (160, 1059), (2337, 1031), (718, 978), (978, 894), (1811, 856), (17406, 841), (837, 829), (369, 732), (805, 711), (529, 643), (141, 639), (642, 637), (2329

**Amazon Book**

**Note: Inside the test.txt of the amazon dataset, there is four user that does not interact with any items**

In [4]:
analyse_dataset("data/amazon-book")

data/amazon-book dataset overview
Total users (user_list)   : 52,643
Total items (item_list)   : 91,599
Train interactions        : 2,380,730
Test  interactions        : 603,378
Train+Test interactions   : 2,984,108
Train sparsity            : 0.000494  (0.049372%)
Test  sparsity            : 0.000125  (0.012513%)

User Coverage
Users in train            : 52,643 / 52,643 (100.000%)
Users in test             : 52,639 / 52,643 (99.992%)

Item Coverage
Items in train: 91,599 / 91,599 (100.000%)
Items in test: 82,629 / 91,599 (90.207%)

Interactions per User Train and Test
  Mean : 45.22, 11.46
  Median  : 26.0, 7.0
  Std     : 77.96, 18.93
  Min     : 16, 1
  Max     : 10682, 2631
Mean interactions per item  : 25.99079
Median : 15.00000
Max (most popular) : 1741
  Top-20 items (id, count) : [(890, 1741), (43750, 1389), (7891, 1244), (36448, 1156), (36471, 1137), (894, 1047), (49138, 1010), (5784, 981), (48171, 927), (1252, 920), (48155, 817), (6830, 813), (36457, 807), (5225, 799), (4881

In [8]:
"""
Script run to fix amazondataset literal issue
Bug fixed: ValueError: invalid literal for int() with base 10: ''
"""
import os
def clean_file(input_path, choice = 1):
    """Clean train.txt / test.txt of amazon-book"""
    output_path = input_path  

    lines_kept = 0
    lines_skipped = 0
    cleaned_lines = []

    with open(input_path, "r", encoding="utf-8") as f:
        for line_num, raw in enumerate(f, 1):
            line = raw.strip()

            if not line:
                lines_skipped += 1
                continue
            tokens = [t for t in line.split(" ") if t != ""]
            if len(tokens) < 2:
                print(f"[SKIP] line {line_num}: only has {len(tokens)} token")
                lines_skipped += 1
                continue

            try:
                [int(t) for t in tokens]
            except ValueError as e:
                print(f"[SKIP] line {line_num}: {e}")
                lines_skipped += 1
                continue

            cleaned_lines.append(" ".join(tokens))
            lines_kept += 1

    if choice == 2:
        print('Delete four single token lines')
        with open(output_path, "w", encoding="utf-8") as f:
            f.write("\n".join(cleaned_lines) + "\n")

    print(f"{input_path}: kept={lines_kept}, skipped={lines_skipped}")


amazon_dir = "data/amazon-book"   

for filename in [ "test.txt", "train.txt" ]:
    path = os.path.join(amazon_dir, filename)
    if os.path.exists(path):
        clean_file(path, 2)
    else:
        print(f"Cannot found: {path}")

[SKIP] line 13648: only has 1 token
[SKIP] line 41590: only has 1 token
[SKIP] line 50737: only has 1 token
[SKIP] line 52235: only has 1 token
Delete four single token lines
data/amazon-book\test.txt: kept=52639, skipped=4
Delete four single token lines
data/amazon-book\train.txt: kept=52643, skipped=0


**Movie Len**

For movielen dataset, the data format is a bit different

The format is a list containing lists, for each item of that list is the item id that the user_id at the current index are interacted with



In [27]:
def format_movielen(dataset_name):
    import json
    if "movielen" not in dataset_name:
        print('Data set should be movielen')
        return
    output_path = dataset_name.replace(".json",".txt")
    output_path = output_path.replace("_data","")
    with open(dataset_name, "r") as f:
        data = json.load(f)
    with open(output_path, 'w') as f:
        for user_id, items in enumerate(data):
            if items:
                line = str(user_id) + ' ' + ' '.join(str(i) for i in items)
                f.write(line + '\n')
    print(f'Conversion {dataset_name} complete')

def analyse_dataset_for_movielen(current_dataset):
    train_users, train_items, train_dict = load_data(f"{current_dataset}/train.txt")
    test_users, test_items, test_dict = load_data(f"{current_dataset}/test.txt")

    # Derive n_users and n_items from union of train + test
    all_users = set(train_users) | set(test_users)
    all_items = set(train_items) | set(test_items)
    n_users = len(all_users)
    n_items = len(all_items)
    n_train = len(train_users)
    n_test  = len(test_users)

    train_users_set = set(train_users)
    test_users_set  = set(test_users)
    train_items_set = set(train_items)
    test_items_set  = set(test_items)

    density_train = n_train / (n_users * n_items)
    density_test  = n_test  / (n_users * n_items)

    print(f"{current_dataset} dataset overview")
    print(f"Total users (union)       : {n_users:,}")
    print(f"Total items (union)       : {n_items:,}")
    print(f"Train interactions        : {n_train:,}")
    print(f"Test  interactions        : {n_test:,}")
    print(f"Train+Test interactions   : {n_train + n_test:,}")
    print(f"Train sparsity            : {density_train:.6f}  ({density_train*100:.6f}%)")
    print(f"Test  sparsity            : {density_test:.6f}  ({density_test*100:.6f}%)")

    print("\nUser Coverage")
    print(f"Users in train            : {len(train_users_set):,} / {n_users:,} ({len(train_users_set)/n_users*100:.3f}%)")
    print(f"Users in test             : {len(test_users_set):,} / {n_users:,} ({len(test_users_set)/n_users*100:.3f}%)")

    print("\nItem Coverage")
    print(f"Items in train            : {len(train_items_set):,} / {n_items:,} ({len(train_items_set)/n_items*100:.3f}%)")
    print(f"Items in test             : {len(test_items_set):,} / {n_items:,} ({len(test_items_set)/n_items*100:.3f}%)")

    train_counts = np.array([len(v) for v in train_dict.values()])
    test_counts  = np.array([len(v) for v in test_dict.values()])
    print("\nInteractions per User (Train / Test)")
    print(f"  Mean   : {train_counts.mean():.2f}  /  {test_counts.mean():.2f}")
    print(f"  Median : {np.median(train_counts):.1f}  /  {np.median(test_counts):.1f}")
    print(f"  Std    : {train_counts.std():.2f}  /  {test_counts.std():.2f}")
    print(f"  Min    : {train_counts.min()}  /  {test_counts.min()}")
    print(f"  Max    : {train_counts.max()}  /  {test_counts.max()}")

    train_item_counts = Counter(train_items)
    test_item_counts  = Counter(test_items)
    train_pop = np.array(list(train_item_counts.values()))
    test_pop  = np.array(list(test_item_counts.values()))

    print(f"\nItem Popularity (Train)")
    print(f"  Mean   : {train_pop.mean():.5f}")
    print(f"  Median : {np.median(train_pop):.5f}")
    print(f"  Max    : {train_pop.max()}")
    print(f"  Top-20 items (id, count) : {train_item_counts.most_common(20)}")

format_movielen("data/movielen/train_data.json")
format_movielen("data/movielen/test_data.json")


Conversion data/movielen/train_data.json complete
Conversion data/movielen/test_data.json complete


In [28]:
analyse_dataset_for_movielen("data/movielen")

data/movielen dataset overview
Total users (union)       : 6,022
Total items (union)       : 3,043
Train interactions        : 796,244
Test  interactions        : 99,455
Train+Test interactions   : 895,699
Train sparsity            : 0.043451  (4.345137%)
Test  sparsity            : 0.005427  (0.542730%)

User Coverage
Users in train            : 6,022 / 6,022 (100.000%)
Users in test             : 6,022 / 6,022 (100.000%)

Item Coverage
Items in train            : 3,043 / 3,043 (100.000%)
Items in test             : 3,021 / 3,043 (99.277%)

Interactions per User (Train / Test)
  Mean   : 132.22  /  16.52
  Median : 76.0  /  10.0
  Std    : 152.98  /  19.12
  Min    : 16  /  2
  Max    : 1720  /  215

Item Popularity (Train)
  Mean   : 261.66415
  Median : 143.00000
  Max    : 2785
  Top-20 items (id, count) : [(1607, 2785), (1375, 2413), (150, 2355), (166, 2282), (2653, 2141), (850, 2135), (2753, 2112), (1350, 2086), (228, 2076), (2756, 2042), (470, 2026), (2769, 2008), (1528, 1992), 

**According to the Hugging Face documentation, the MovieLens-1M dataset consists of 1,000,209 anonymous ratings covering approximately 3,900 movies from 6,040 users. Following the preprocessing protocol used in the LCF framework, the data are split into training, validation, and test sets. In our implementation, only the training and test splits are considered, which results in 6,022 users and 3,043 items, as the validation data are excluded.**


